store/tikv/store_aliveness.go

```diff
@@ -0,0 +1,272 @@
+
+func newStoreLiveness(pdHosts []string, timeout time.Duration) *StoreLiveness {
+	if len(pdHosts) == 0 {
+		return nil
+	}
+	doneCh := make(chan struct{})
+	l := &StoreLiveness{
+		timeout:        timeout,
+		statusResolver: newStatusAddressResolver(pdHosts, doneCh),
+		doneCh:         doneCh,
+	}
+	return l
+}
+
+// IsAlive checks whether store process still alive.
+func (s *StoreLiveness) IsAlive(storeID uint64) bool {
+	if s == nil {
+		return false
+	}
+	_, err, _ := s.sf.Do(strconv.FormatUint(storeID, 10), func() (interface{}, error) {
+		err := s.checkKvStatusAPI(storeID)
+		if err != nil {
+			logutil.BgLogger().Warn("[liveness] send fail and store is down, need invalid region cache and switch next peer",
+				zap.Uint64("store", storeID), zap.Error(err))
+			return nil, err
+		}
+		return nil, nil
+	})
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+func (s *StoreLiveness) start() {
+	failpoint.Inject(knobSwitch, func() {
+		if s.statusResolver.knobs.mockStart != nil {
+			s.statusResolver.knobs.mockStart(s.statusResolver)
+		}
+	})
+	s.statusResolver.start()
+}
+
+func (s *StoreLiveness) close() {
+	if s == nil {
+		return
+	}
+	close(s.doneCh)
+}
+
+func (s *StoreLiveness) checkKvStatusAPI(storeID uint64) error {
+	statusAddr := s.statusResolver.resolve(storeID)
+	if len(statusAddr) == 0 {
+		return errors.Errorf("unknown store %d maybe caused by offline in pd or cache outdated", storeID)
+	}
+	failpoint.Inject(knobSwitch, func() {
+		if s.knobs.mockTiKVStatus != nil {
+			failpoint.Return(s.knobs.mockTiKVStatus(statusAddr))
+		}
+	})
+	ctx, cancel := context.WithTimeout(context.Background(), s.timeout)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, statusAddr, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := util.InternalHTTPClient().Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err1 := resp.Body.Close()
+		if err1 != nil {
+			logutil.BgLogger().Error("close kv status api body failed", zap.Error(err))
+		}
+	}()
+	if resp.StatusCode != http.StatusOK {
+		return errors.Errorf("request kv status fail, ret: %s", resp.Status)
+	}
+	return nil
+}
+
+type kvStatusAddrResolver struct {
+	doneCh          <-chan struct{}
+	storeStatusAddr unsafe.Pointer //map[uint64]string
+	pdHosts         []string
+	pdIdx           int
+	knobs           struct {
+		mockFetchStoreStatus func(url string) (storesStat storesState, err error)
+		mockStart            func(r *kvStatusAddrResolver)
+	}
+}
+
+func newStatusAddressResolver(pdHosts []string, doneCh chan struct{}) *kvStatusAddrResolver {
+	saMap := make(map[string]string)
+	r := &kvStatusAddrResolver{
+		storeStatusAddr: unsafe.Pointer(&saMap),
+		pdHosts:         pdHosts,
+		doneCh:          doneCh,
+	}
+	return r
+}
+
+func (r *kvStatusAddrResolver) start() {
+	err := r.refreshKvStatusAddrForAnyPD()
+	if err != nil {
+		logutil.BgLogger().Warn("all pd is down at startup, store-liveness is disabled and will works after pd recovery")
```

Comment:
How the store-liveness is disabled?
There is no `disable/enable` flag.

Reply:
`storeLiveness#IsAlive` will return false when it can not know kv status port just like "store-liveness function" be disabled. :D

and after pd recovered,  storeLiveness#IsAlive will come back to work immediately.